File: 1. Spanish airports/SpanishAirports.java
package org.jlpuente.RDD;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

/**
 * RDD-based Java Spark program that reads a CSV file,
 * which must be in the root directory of the project,
 * and counts the number of Spanish airports by type of airport.
 */

public class SpanishAirports {

    public static void main(String[] args) {

        // Step 0. Turn off the log messages
        Logger.getRootLogger().setLevel(Level.OFF);

        // Step 1. Create a SparkConf object
        SparkConf sparkConf = new SparkConf()
                .setMaster("local[4]")
                .setAppName("SpanishAirports");

        // Step 2. Create a Java Spark Context
        JavaSparkContext sparkContext = new JavaSparkContext(sparkConf);

        // Step 3. Point to the CSV file
        JavaRDD<String> lines = sparkContext.textFile("airports.csv");

        // Step 4. Read the header
        String header = lines.first();

        // Step 4.1. Store the indices and the Spanish ISO code
        int iso_country_index = 8;
        int name_index = 2;
        String spanish_ISO_code = "ES";

        // Step 5. Make the stream
        JavaPairRDD<Integer, String> spanishAiportsType = lines
                .filter((String s) -> !s.contains(header))
                .mapToPair(airport -> new Tuple2<>(airport.split(",")[iso_country_index],
                        airport.split(",")[name_index]))
                .filter(country -> country._1.contains(spanish_ISO_code))
                .mapToPair(pair -> new Tuple2<>(pair._2(), 1))
                .reduceByKey(Integer::sum)
                .mapToPair(pair -> new Tuple2<>(pair._2(), pair._1()))
                .sortByKey(false)
                .cache();

        // Step 6. Print the results on screen
        spanishAiportsType.collect().forEach(type -> System.out.println(type._2 + ": " +  type._1));

        // Step 7. Save results as text file
        spanishAiportsType.saveAsTextFile("output_spanish_airports");

        // Step 8. Close the SparkContext object to release the resources allocated
        sparkContext.close();

    }
}

File: 1. Spanish airports/SpanishAirports.py
from pyspark import SparkContext, SparkConf
import csv

'''
RDD-based PySpark program that reads a CSV file,
which must be in the root directory of the project,
and counts the number of Spanish airports by type of airport.
'''


def main() -> None:

    # Step 1. Create a SparkConf object
    spark_conf = SparkConf().setAppName("SpanishAirportsCounter").setMaster("local[4]")

    # Step 2. Create a SparkContext object
    spark_context = SparkContext(conf=spark_conf)

    # Step 3. Read the CSV file
    lines = spark_context.textFile("airports.csv")

    # Step 3.1. Store Spanish ISO country code and header indices
    spanish_iso_code = "ES"
    iso_country_index = 8
    airport_type_index = 2

    # Step 4. Parse the CSV data and filter for Spanish airports by using csv Python module
    spanish_airports_rdd = lines \
        .map(lambda line: next(csv.reader([line]))) \
        .filter(lambda fields: fields[iso_country_index] == spanish_iso_code)

    # Step 5. Count the number of Spanish airports by type of airport
    airport_type_counts = spanish_airports_rdd \
        .map(lambda fields: (fields[airport_type_index], 1)) \
        .reduceByKey(lambda x, y: x + y)

    # Step 6. Print the output
    for airport_type, count in airport_type_counts.collect():
        print(f'{airport_type}: {count}')

    # Step 7. Save output at text file
    airport_type_counts.saveAsTextFile("output_spanish_airports_pyspark")

    # Stop the SparkContext
    spark_context.stop()


if __name__ == "__main__":

    main()
